Fixed #9475 -- Allowed RelatedManager.add(), create(), etc. for m2m with a through model.

collinanderson opened the pull request:
https://code.djangoproject.com/ticket/9475

claudep commented:
Very promising!
Do we have a test where the through table requires values and `through_defaults` is empty?

collinanderson commented:
Ok I added some docs and added a test for a case where the through table requires a value.

airstandley commented:
Shouldn't line 955 be `self.add(*objs, through_defaults=through_defaults)`?

collinanderson commented:
@airstandley Nice catch! Fixed.

collinanderson commented:
I just rebased this patch, updated version numbers, and added fail and success tests for add, create, set, get_or_create, and update_or_create. The failed mariadb test looks like a fluke.

timgraham commented:
How about some helpful message if `through_defaults` is required but not specified (or specified insufficiently or incorrectly)?

claudep commented:
Thanks Collin! This would be a nice addition for 2.2, hopefully we can commit it before feature freeze, even if we have to polish some aspects during the alpha phase.

collinanderson commented:
Hmm... I don't think it's possible to detect before the fact if through_defaults is needed, because the field could get auto filled in in the save() method, so basically the only way to know if it will work is to catch the error. But, it seems wrong to me to catch the `IntegrityError` and then raise a different error.

I'm hoping that letting the error message `IntegrityError: NOT NULL constraint failed: myapp_mymodel.myrequiredfield` bubble up is clear enough.

But maybe catching and raising a new error is helpful? Do you think something like the code below?

```
try:
    self.through._default_manager.using(db).bulk_create(**etc)
except IntegrityError as e:
    if rel.through._meta.auto_created:
        raise
    raise IntegrityError("Try specifying through_defaults or giving your field a default value")
```

collinanderson commented:
Ohh, I see now we're using `bulk_create` so values can't be filled in using a custom `save()` method...

timgraham commented on tests/m2m_through_regress/tests.py:
```diff
@@ -47,41 +47,17 @@ def test_retrieve_forward_m2m_items(self):
-        with self.assertRaisesMessage(AttributeError, msg):
-            self.bob.group_set.set([])
+    def test_set_on_reverse_m2m_with_intermediary_model(self):
+        self.bob.group_set.set([])
```
Are these tests checking something that the `m2m_through` tests aren't? It looks odd that the results of these calls aren't verified.

collinanderson replied:
Yeah, good point. I was mostly just converting "can't" test into "can" tests. Should I just delete them these?

timgraham replied:
I didn't look closely but I don't think they add anything (but please audit test coverage as I said in my other comment).

collinanderson replied:
Ok, I removed these tests.

collinanderson commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -1022,7 +1002,7 @@ def set(self, objs, *, clear=False):
                 if clear:
                     self.clear()
-                    self.add(*objs)
+                    self.add(*objs, through_defaults=through_defaults)
```
Ok, I just added tests for this.

pope1ni commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -956,32 +956,23 @@ def count(self):
+                # If this is a symmetrical m2m relation to self, add the mirror
+                # entry in the m2m table. `through_defaults` aren't used here
+                # because of the system check error fields.E332: Many-to-many
+                # fields with intermediate tables mus not be symmetrical.
```
`mus` → `must`.

pope1ni commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -1130,10 +1105,10 @@ def _add_items(self, source_field_name, target_field_name, *objs):
                     # Add the ones that aren't there already
                     self.through._default_manager.using(db).bulk_create([
-                        self.through(**{
+                        self.through(**dict(through_defaults, **{
```
Can replace `**dict(through_defaults, **{…})` and use unpacking generalisations: `**{**through_defaults, **{…}}`.

charettes replied:
I guess `**through_defaults, **{…}` would work as well given surfacing a `TypeError` if any of the ids properties are provided through `through_defaults` should probably not go unnoticed.

pope1ni commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -956,32 +956,23 @@ def count(self):
-                    "intermediary model. Use %s.%s's Manager instead." %
-                    (opts.app_label, opts.object_name)
-                )
+        def add(self, *objs, through_defaults=None):
```
I noticed that we use keyword-only arguments for `def set(self, objs, *, clear=False, through_defaults=None):`

Do we want to so something similar for other methods:
- `def add(self, *objs, *, through_defaults=None)`:
- `def create(self, *, through_defaults=None, **kwargs):`
- `def get_or_create(self, *, through_defaults=None, **kwargs):`
- `def update_or_create(self, *, through_defaults=None, **kwargs):`

timgraham replied:
Sure, I will push those change, except the suggestion for `add()` gives invalid syntax.

pope1ni replied:
Oh. Strange. Not sure why that doesn't work. Thanks.

charettes commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -1130,10 +1105,10 @@ def _add_items(self, source_field_name, target_field_name, *objs):
-                        self.through(**{
+                        self.through(**dict(through_defaults, **{
                             '%s_id' % source_field_name: self.related_val[0],
                             '%s_id' % target_field_name: obj_id,
```
Also the `%s_id` is probably going to break for fields that define different `db_column`s. This wasn't an issue before because this branch could only be reached with auto-created intemediary models but now that we allow all forms of `through` we should use `attname` or directly `*_name` here.

charettes replied:
My assumptions here seems to be wrong, this should work fine.

https://github.com/django/django/blob/d212bc03bac1bce074207074a45032bb6bfdf3ec/django/db/models/fields/related.py#L916-L917
